File: lib/ace/adaptation.ts
import { 
  Sample, 
  EnvironmentResult, 
  AdapterStepResult, 
  TaskEnvironment, 
  GeneratorOutput, 
  Playbook 
} from './types'
import { Generator, Reflector, Curator } from './roles'
import { PlaybookImpl } from './playbook'

export class GTVTaskEnvironment implements TaskEnvironment {
  async evaluate(sample: Sample, generator_output: GeneratorOutput): Promise<EnvironmentResult> {
    // 简化的GTV评估逻辑
    // 实际实现应该调用真实的GTV评估API
    const isCorrect = this.evaluateGTVResponse(sample, generator_output)
    const feedback = isCorrect 
      ? "回答正确，符合GTV签证要求" 
      : "回答需要改进，请参考GTV签证官方指南"
    
    return {
      feedback,
      ground_truth: sample.ground_truth,
      metrics: {
        correctness: isCorrect ? 1 : 0,
        completeness: this.calculateCompleteness(generator_output),
        relevance: this.calculateRelevance(sample, generator_output)
      }
    }
  }

  private evaluateGTVResponse(sample: Sample, output: GeneratorOutput): boolean {
    // 简化的正确性评估
    // 实际应该使用更复杂的GTV专业知识评估
    const answer = output.final_answer.toLowerCase()
    const question = sample.question.toLowerCase()
    
    // 基本关键词匹配
    if (question.includes('gtv') && answer.includes('global talent visa')) {
      return true
    }
    if (question.includes('eligibility') && answer.includes('criteria')) {
      return true
    }
    if (question.includes('application') && answer.includes('process')) {
      return true
    }
    
    return false
  }

  private calculateCompleteness(output: GeneratorOutput): number {
    // 基于回答长度和内容丰富度计算完整性
    const reasoningLength = output.reasoning.length
    const answerLength = output.final_answer.length
    const bulletCount = output.bullet_ids.length
    
    // 归一化到0-1范围
    const reasoningScore = Math.min(reasoningLength / 500, 1)
    const answerScore = Math.min(answerLength / 200, 1)
    const bulletScore = Math.min(bulletCount / 5, 1)
    
    return (reasoningScore + answerScore + bulletScore) / 3
  }

  private calculateRelevance(sample: Sample, output: GeneratorOutput): number {
    // 基于关键词匹配计算相关性
    const questionWords = sample.question.toLowerCase().split(/\s+/)
    const answerWords = output.final_answer.toLowerCase().split(/\s+/)
    
    const commonWords = questionWords.filter(word => 
      answerWords.includes(word) && word.length > 3
    )
    
    return Math.min(commonWords.length / questionWords.length, 1)
  }
}

export class ACEAdapter {
  private playbook: PlaybookImpl
  private recentReflections: string[] = []
  private reflectionWindow = 3

  constructor(
    private generator: Generator,
    private reflector: Reflector,
    private curator: Curator,
    initialPlaybook?: PlaybookImpl
  ) {
    this.playbook = initialPlaybook || new PlaybookImpl()
  }

  async processSample(
    sample: Sample,
    environment: TaskEnvironment,
    progress: string
  ): Promise<AdapterStepResult> {
    // 1. 生成回答
    const generatorOutput = await this.generator.generate({
      question: sample.question,
      context: sample.context,
      playbook: this.playbook,
      reflection: this.getReflectionContext()
    })

    // 2. 环境评估
    const environmentResult = await environment.evaluate(sample, generatorOutput)

    // 3. 反思
    const reflection = await this.reflector.reflect({
      question: sample.question,
      generator_output: generatorOutput,
      playbook: this.playbook,
      ground_truth: environmentResult.ground_truth,
      feedback: environmentResult.feedback
    })

    // 4. 应用反思标签
    this.applyBulletTags(reflection)
    this.updateRecentReflections(reflection)

    // 5. 策展更新
    const curatorOutput = await this.curator.curate({
      reflection,
      playbook: this.playbook,
      question_context: this.buildQuestionContext(sample, environmentResult),
      progress
    })

    // 6. 应用知识库更新
    this.playbook.apply_delta(curatorOutput.delta)

    return {
      sample,
      generator_output: generatorOutput,
      environment_result: environmentResult,
      reflection,
      curator_output: curatorOutput,
      playbook_snapshot: this.playbook.as_prompt()
    }
  }

  private getReflectionContext(): string {
    return this.recentReflections.join('\n---\n')
  }

  private updateRecentReflections(reflection: any): void {
    const serialized = JSON.stringify(reflection.raw)
    this.recentReflections.push(serialized)
    
    if (this.recentReflections.length > this.reflectionWindow) {
      this.recentReflections = this.recentReflections.slice(-this.reflectionWindow)
    }
  }

  private applyBulletTags(reflection: any): void {
    if (reflection.bullet_tags && Array.isArray(reflection.bullet_tags)) {
      reflection.bullet_tags.forEach((tag: any) => {
        try {
          this.playbook.tag_bullet(tag.id, tag.tag)
        } catch (error) {
          // 忽略标签错误
        }
      })
    }
  }

  private buildQuestionContext(sample: Sample, environmentResult: EnvironmentResult): string {
    return [
      `question: ${sample.question}`,
      `context: ${sample.context}`,
      `metadata: ${JSON.stringify(sample.metadata)}`,
      `feedback: ${environmentResult.feedback}`,
      `ground_truth: ${environmentResult.ground_truth || 'none'}`
    ].join('\n')
  }

  getPlaybook(): PlaybookImpl {
    return this.playbook
  }

  getStats() {
    return this.playbook.stats()
  }
}

File: ACE-open/ace/adaptation.py
"""Adaptation loops for offline and online ACE training."""

from __future__ import annotations

import json
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Dict, Iterable, List, Optional, Sequence

from .playbook import Playbook
from .roles import Curator, CuratorOutput, Generator, GeneratorOutput, Reflector, ReflectorOutput


@dataclass
class Sample:
    """Single task instance presented to ACE."""

    question: str
    context: str = ""
    ground_truth: Optional[str] = None
    metadata: Dict[str, object] = field(default_factory=dict)


@dataclass
class EnvironmentResult:
    """Feedback returned by the task environment after executing the generator output."""

    feedback: str
    ground_truth: Optional[str]
    metrics: Dict[str, float] = field(default_factory=dict)


class TaskEnvironment(ABC):
    """Defines how to evaluate generator outputs for a sample."""

    @abstractmethod
    def evaluate(
        self, sample: Sample, generator_output: GeneratorOutput
    ) -> EnvironmentResult:
        """Return environment feedback plus optional ground truth information."""


@dataclass
class AdapterStepResult:
    sample: Sample
    generator_output: GeneratorOutput
    environment_result: EnvironmentResult
    reflection: ReflectorOutput
    curator_output: CuratorOutput
    playbook_snapshot: str


class AdapterBase:
    """Shared orchestration logic for offline and online ACE adaptation."""

    def __init__(
        self,
        *,
        playbook: Optional[Playbook] = None,
        generator: Generator,
        reflector: Reflector,
        curator: Curator,
        max_refinement_rounds: int = 1,
        reflection_window: int = 3,
    ) -> None:
        self.playbook = playbook or Playbook()
        self.generator = generator
        self.reflector = reflector
        self.curator = curator
        self.max_refinement_rounds = max_refinement_rounds
        self.reflection_window = reflection_window
        self._recent_reflections: List[str] = []

    # ------------------------------------------------------------------ #
    def _reflection_context(self) -> str:
        return "\n---\n".join(self._recent_reflections)

    def _update_recent_reflections(self, reflection: ReflectorOutput) -> None:
        serialized = json.dumps(reflection.raw, ensure_ascii=False)
        self._recent_reflections.append(serialized)
        if len(self._recent_reflections) > self.reflection_window:
            self._recent_reflections = self._recent_reflections[-self.reflection_window :]

    def _apply_bullet_tags(self, reflection: ReflectorOutput) -> None:
        for tag in reflection.bullet_tags:
            try:
                self.playbook.tag_bullet(tag.id, tag.tag)
            except ValueError:
                continue

    def _question_context(self, sample: Sample, environment_result: EnvironmentResult) -> str:
        parts = [
            f"question: {sample.question}",
            f"context: {sample.context}",
            f"metadata: {json.dumps(sample.metadata)}",
            f"feedback: {environment_result.feedback}",
            f"ground_truth: {environment_result.ground_truth}",
        ]
        return "\n".join(parts)

    def _progress_string(self, epoch: int, total_epochs: int, step: int, total_steps: int) -> str:
        return f"epoch {epoch}/{total_epochs} · sample {step}/{total_steps}"

    def _process_sample(
        self,
        sample: Sample,
        environment: TaskEnvironment,
        *,
        epoch: int,
        total_epochs: int,
        step_index: int,
        total_steps: int,
    ) -> AdapterStepResult:
        generator_output = self.generator.generate(
            question=sample.question,
            context=sample.context,
            playbook=self.playbook,
            reflection=self._reflection_context(),
        )
        env_result = environment.evaluate(sample, generator_output)
        reflection = self.reflector.reflect(
            question=sample.question,
            generator_output=generator_output,
            playbook=self.playbook,
            ground_truth=env_result.ground_truth,
            feedback=env_result.feedback,
            max_refinement_rounds=self.max_refinement_rounds,
        )
        self._apply_bullet_tags(reflection)
        self._update_recent_reflections(reflection)
        curator_output = self.curator.curate(
            reflection=reflection,
            playbook=self.playbook,
            question_context=self._question_context(sample, env_result),
            progress=self._progress_string(epoch, total_epochs, step_index, total_steps),
        )
        self.playbook.apply_delta(curator_output.delta)
        return AdapterStepResult(
            sample=sample,
            generator_output=generator_output,
            environment_result=env_result,
            reflection=reflection,
            curator_output=curator_output,
            playbook_snapshot=self.playbook.as_prompt(),
        )


class OfflineAdapter(AdapterBase):
    """Runs multi-epoch offline adaptation on a training split."""

    def run(
        self,
        samples: Sequence[Sample],
        environment: TaskEnvironment,
        epochs: int = 1,
    ) -> List[AdapterStepResult]:
        results: List[AdapterStepResult] = []
        total_steps = len(samples)
        for epoch_idx in range(1, epochs + 1):
            for step_idx, sample in enumerate(samples, start=1):
                result = self._process_sample(
                    sample,
                    environment,
                    epoch=epoch_idx,
                    total_epochs=epochs,
                    step_index=step_idx,
                    total_steps=total_steps,
                )
                results.append(result)
        return results


class OnlineAdapter(AdapterBase):
    """Processes a stream of samples sequentially, updating the playbook in-place."""

    def run(
        self,
        samples: Iterable[Sample],
        environment: TaskEnvironment,
    ) -> List[AdapterStepResult]:
        results: List[AdapterStepResult] = []
        step_idx = 0
        for step_idx, sample in enumerate(samples, start=1):
            result = self._process_sample(
                sample,
                environment,
                epoch=1,
                total_epochs=1,
                step_index=step_idx,
                total_steps=step_idx,
            )
            results.append(result)
        return results
